configuration/outbound.md

```diff
@@ -0,0 +1,201 @@
+# Outbound push [beta]
+
+*Outbound is only available in Medic 3.5.0 and above*
+
+Outbound push allows configurers to have single reports generated by CHWs or others trigger outbound REST requests using the data in that report. For example, upon receiving a referral report you could send that referral to an external facility system that will manage and process that event.
+
+This feature is experimental, and may be replaced by some third party technology over time as we discover the feature-set we're looking for.
+
+## Configuration
+
+For outbound pushes to occur, you must [enable the `mark_for_outbound` transition in config](./transitions.md#Configurations):
+```json
+{
+  "transitions": {
+    "mark_for_outbound": true
+  }
+}
+```
+
+The rest of the configuration is against the `outbound` configuration property, which allows you to configure as many outbound pushes as you like, keyed off a unique name:
+
+```json
+{
+  "outbound": {
+    "first config": {
+      "relevant_to": "...",
+      "destination": {},
+      "mapping": {}
+    },
+    "second config": { }
+  }
+}
+```
+
+Each outbound push configuration contains the following properties: `relevant_to`, `destination` and `mapping`.
+
+### relevant_to
+
+An "expression" (some JS code that resolves to a truthy or falsy value) that determines whether this configuration is relevant to the document that is being tested. The document is passed to the expression is `doc`, and is fully hydrated before being passed.
+
+Example: you want to send a referral to an external facility tool when a CHW refers a patient:
+
+```json
+{
+  "relevant_to": "doc.type === 'data_record' && doc.form === 'referral'"
+}
+```
+
+### destination
+
+A complex property that declares how the data should be passed to the external service. It currently supports two authentication types, Basic Auth and a custom authentication mode for Muso SIH.
+
+Example:
+
+```json
+{
+  "destination": {
+    "base_url": "https://example.com",
+    "auth": {
+      "type": "basic",
+      "username": "admin",
+      "password_key": "example.com"
+    },
+    "path": "/api/v1/referral"
+  }
+}
+```
+
+Here `password_key` is a key used to find the password in CouchDB's node-based configuration. See Credentials section below.
+
+If you don't provide an authentication parameter then the request will be sent without authentication.
+
+### mapping
+
+A complex property that declares how the payload to be sent to the `destination` should be created.
+
+Each key is a JSON path to a location in the payload, and each value is either:
```

Comment:
JSON path is a named thing (there is a `json-path` library for example), but it's not really correct as it's is basically like what jq uses, which is not what we support at all.

However, we don't exactly support object-path's api either: we basically just support whatever you'd call accessing a deep parameter in JS e.g., `doc.field.blah`. What would you call that?

Reply:
> we don't exactly support object-path's api either

What do you mean by this?

Reply:
> What do you mean by this exactly?

I mean the explanation directly after the text you're quoting, that we only really support a single string that represents a path to deeper properties. To look at the linked documentation of yours, the only thing we support is the very first get example and none of the others.

> Is this the library you're talking about? json-path/JsonPath. **I'm not sure we support this syntax..**. for example $.store.book[0].title from their documentation won't work in our system? I think we would have store.book.0.title

Yes, in my comment I said we don't support the syntax, I think you're agreeing with me?

Reply:
So one option is if we could find out what this kind of path traversal is called traditionally / globally in the JSON community (if it exists) we could link to documentation about just that.

Otherwise we could have a little section about its semantics?

Reply:
On second thought I just linked to the usage section and renamed everything from JSON path to object path.